flink-java/src/main/java/org/apache/flink/api/java/utils/PlanGenerator.java

```diff
@@ -0,0 +1,126 @@
+import org.apache.flink.api.java.operators.DataSink;
+import org.apache.flink.api.java.operators.OperatorTranslation;
+import org.apache.flink.api.java.typeutils.runtime.kryo.Serializers;
+import org.apache.flink.util.Visitor;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * A generator that generates a {@link Plan} from a graph of {@link Operator}s.
+ */
+public class PlanGenerator {
+	private static final Logger LOG = LoggerFactory.getLogger(PlanGenerator.class);
+
+	private final List<DataSink<?>> sinks;
+	private final ExecutionConfig config;
+	private final String jobName;
+
+	public PlanGenerator(
+			List<DataSink<?>> sinks,
+			ExecutionConfig config,
+			String jobName) {
+		this.sinks = sinks;
+		this.config = config;
+		this.jobName = jobName;
+	}
+
+	public Plan generate() {
+		OperatorTranslation translator = new OperatorTranslation();
+		Plan plan = translator.translateToPlan(sinks, jobName);
+
+		if (config.getParallelism() > 0) {
+			plan.setDefaultParallelism(config.getParallelism());
+		}
+		plan.setExecutionConfig(config);
+
+		// Check plan for GenericTypeInfo's and register the types at the serializers.
+		if (!config.isAutoTypeRegistrationDisabled()) {
+			plan.accept(new Visitor<Operator<?>>() {
+
+				private final Set<Class<?>> registeredTypes = new HashSet<>();
+				private final Set<org.apache.flink.api.common.operators.Operator<?>> visitedOperators = new HashSet<>();
+
+				@Override
+				public boolean preVisit(org.apache.flink.api.common.operators.Operator<?> visitable) {
+					if (!visitedOperators.add(visitable)) {
+						return false;
+					}
+					OperatorInformation<?> opInfo = visitable.getOperatorInfo();
+					Serializers.recursivelyRegisterType(opInfo.getOutputType(), config, registeredTypes);
+					return true;
+				}
+
+				@Override
+				public void postVisit(org.apache.flink.api.common.operators.Operator<?> visitable) {
+				}
+			});
+		}
+
+		// All types are registered now. Print information.
+		int registeredTypes = config.getRegisteredKryoTypes().size() +
+				config.getRegisteredPojoTypes().size() +
+				config.getRegisteredTypesWithKryoSerializerClasses().size() +
+				config.getRegisteredTypesWithKryoSerializers().size();
+		int defaultKryoSerializers = config.getDefaultKryoSerializers().size() +
+				config.getDefaultKryoSerializerClasses().size();
+		LOG.info("The job has {} registered types and {} default Kryo serializers", registeredTypes,
+				defaultKryoSerializers);
+
+		if (config.isForceKryoEnabled() && config.isForceAvroEnabled()) {
+			LOG.warn("In the ExecutionConfig, both Avro and Kryo are enforced. Using Kryo serializer");
+		}
+		if (config.isForceKryoEnabled()) {
+			LOG.info("Using KryoSerializer for serializing POJOs");
+		}
+		if (config.isForceAvroEnabled()) {
+			LOG.info("Using AvroSerializer for serializing POJOs");
+		}
```

Comment:
This block here seems a bit weird, right? It may end up printing:
```
LOG.warn("In the ExecutionConfig, both Avro and Kryo are enforced. Using Kryo serializer");
LOG.info("Using KryoSerializer for serializing POJOs");
LOG.info("Using AvroSerializer for serializing POJOs");
```
which is misleading as either `kryo` or `avro` will be used.

Reply:
Thank for detailed check, I just copy this from `ExecutionEnvironment#createProgramPlan` method, I will correct this.